trillian/ctdns/handlers.go

```diff
@@ -0,0 +1,311 @@
+// https://github.com/google/certificate-transparency-rfcs/blob/master/dns/draft-ct-over-dns.md
+
+package ctdns
+
+import (
+	"context"
+	"encoding/base32"
+	"encoding/base64"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/golang/glog"
+	"github.com/google/certificate-transparency-go/trillian/ctfe"
+	"github.com/google/certificate-transparency-go/trillian/ctfe/configpb"
+	"github.com/google/trillian"
+	"github.com/google/trillian/types"
+	"github.com/miekg/dns"
+)
+
+// The maximum length of the text returned in a DNS consistency or inclusion
+// proof response.
+const maxProofLen = 255
+
+var (
+	// In these formats the last (.*) is always the zone, this is checked outside
+	// the regex matching.
+	sthRE     = regexp.MustCompile("^sth\\.(.*)\\.$")
+	consistRE = regexp.MustCompile("^(\\d+)\\.(\\d+)\\.(\\d+)\\.sth-consistency\\.(.*)\\.$")
+	hashRE    = regexp.MustCompile("^([A-Z0-9]+)\\.hash\\.(.*)\\.$")
+	treeRE    = regexp.MustCompile("^(\\d+)\\.(\\d+)\\.(\\d+)\\.tree\\.(.*)\\.$")
+)
+
+type dnsFunc func(*CTDNSHandler, []string, dns.ResponseWriter, *dns.Msg)
+
+type CTDNSHandler struct {
+	client   trillian.TrillianLogClient
+	cfg      *configpb.LogConfig
+	opts     ctfe.InstanceOptions
+	handlers []dnsHandler
+}
+
+type dnsHandler struct {
+	matchRE  *regexp.Regexp
+	handleFn dnsFunc
+}
+
+func New(client trillian.TrillianLogClient, cfg *configpb.LogConfig, opts ctfe.InstanceOptions) dns.Handler {
+	return &CTDNSHandler{
+		client: client,
+		cfg:    cfg,
+		opts:   opts,
+		handlers: []dnsHandler{
+			{matchRE: sthRE, handleFn: sthFunc},
+			{matchRE: consistRE, handleFn: consistFunc},
+			{matchRE: hashRE, handleFn: hashFunc},
+			{matchRE: treeRE, handleFn: treeFunc},
+		},
+	}
+}
+
+func (c *CTDNSHandler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	// From the spec all requests must be class INET, type TXT.
+	if err := validate(r); err != nil {
+		glog.Warningf("CTDNSHandler.ServeDNS(): %v", err)
+		failWithRcode(w, r, dns.RcodeFormatError)
+		return
+	}
+
+	glog.V(1).Infof("CTDNSHandler.ServeDNS(): Query: %v", r.Question[0].Name)
+	q := dns.Fqdn(r.Question[0].Name)
+
+	// Find the first handler that accepts the name pattern match.
+	for _, h := range c.handlers {
+		params := h.matchRE.FindStringSubmatch(q)
+		// All our matches should include at least the zone as a capture group
+		// so we must have at least two results from the above.
+		if len(params) > 1 {
+			// Additionally check that the zone matched the last regex param and the
+			// whole string matched to avoid any false positives.
+			if params[0] == q && params[len(params)-1] == c.cfg.DnsZone {
+				// This handler accepted the match and provides the result.
+				h.handleFn(c, params, w, r)
+				return
+			}
+		}
+	}
+	// No handler matched. Reject the request.
+	failWithRcode(w, r, dns.RcodeNotZone)
+}
+
+func sthFunc(c *CTDNSHandler, params []string, w dns.ResponseWriter, r *dns.Msg) {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), c.opts.Deadline)
```

Comment:
Can we move this up into `ServeDNS` and have all of the handlers take an (already-time-constrained) context.

Reply:
Done.